parser/charset/encoding.go

```diff
@@ -140,9 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
 	return dest
 }
 
+// validUTF8 checks whether there are illegal utf8 characters in []byte.
+func (e *Encoding) validUTF8(src []byte) ([]byte, error) {
+	resultBytes := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRune(src)
+		if r == utf8.RuneError && size == 1 {
+			return resultBytes, e.generateErr(src, 3)
+		}
+		src = src[size:]
+	}
+
+	return resultBytes, nil
+}
+
+// validUTF8String checks whether there are illegal utf8 characters in string.
+func (e *Encoding) validUTF8String(src string) (string, error) {
+	resultStr := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRuneInString(src)
+		if r == utf8.RuneError && size == 1 {
+			return resultStr, e.generateErr([]byte(src), 3)
+		}
+		src = src[size:]
+	}
+
+	return resultStr, nil
+}
+
 // Decode convert bytes from a specific charset to utf-8 charset.
 func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
 	if !e.enabled() {
+		if e.name == "utf-8" {
+			return e.validUTF8(src)
```

Comment:
Why do we need to check utf8 here?

Reply:
I got it. But it seems that only `char` using this check?

Reply:
If we have some invalid `utf8` characters in store, add this check could find it.

Reply:
Well, I mean we already do check on the parser layer, and we'll do it repeatedly when we call it.

Reply:
Create a new PR to follow the invlid UTF8 character in buitin func #29685. And delete the UTF8 check in this PR. @zimulala